algorithmic_efficiency/logging_utils.py

```diff
@@ -0,0 +1,492 @@
+    measurements['goal_reached'] = goal_reached
+    measurements['is_time_remaining'] = is_time_remaining
+    measurements['early_stop'] = early_stop
+    measurements['training_complete'] = training_complete
+
+    # Record hyperparameters
+    measurements['batch_size'] = batch_size
+    if hyperparameters:
+      hparams_dict = hyperparameters._asdict()
+      # prefix every key with "hparam." to make more human-readable and to
+      # avoid overlap with other possible kys in the measurements dict.
+      hparams_dict = {f'hparam.{k}': v for k, v in hparams_dict.items()}
+      measurements.update(hparams_dict)
+
+    # Record workload properties
+    workload_properties = _get_workload_properties(workload)
+    measurements.update(workload_properties)
+
+    if FLAGS.extra_metadata:
+      extra_metadata = _get_extra_metadata_as_dict(FLAGS.extra_metadata)
+      measurements.update(extra_metadata)
+
+    # Record utilization
+    utilization_measurements = _get_utilization()
+    measurements.update(utilization_measurements)
+
+    return measurements
+
+  def save_eval(self, workload: spec.Workload,
+                hyperparameters: Optional[spec.Hyperparamters], trial_idx: int,
+                global_step: int, batch_size: int, latest_eval_result: dict,
+                global_start_time: float, accumulated_submission_time: float,
+                goal_reached: bool, is_time_remaining: bool,
+                training_complete: bool, early_stop: bool):
+    """"Write or append to "measurements.csv".
+
+    A "measurements.csv" is created for each hyperparameter tuning trial and a
+    row is appended for every model evaluation. The information included is
+    loss, accuracy, training step, time elapsed, hparams, workload properties,
+    and hardware utilization."""
+    measurements = self._get_eval_measurements(
+        workload, hyperparameters, trial_idx, global_step, batch_size,
+        latest_eval_result, global_start_time, accumulated_submission_time,
+        goal_reached, is_time_remaining, training_complete, early_stop)
+
+    # Save to CSV file
+    trial_output_path = os.path.join(self.workload_log_dir,
+                                     'trial_' + str(trial_idx))
+    os.makedirs(trial_output_path, exist_ok=True)
+    csv_path = os.path.join(trial_output_path, 'measurements.csv')
+    logging.info(f'Recording measurements to: {csv_path}')
+    self._append_to_csv(measurements, csv_path)
+
+  def check_eval_frequency_override(self, workload: spec.Workload,
+                                    global_step: int, batch_size: int):
+    """Parse the eval_frequency_override cli argument and return whether or not
+    the user wants to eval this step."""
+    if not FLAGS.eval_frequency_override:
+      return False
+
+    try:
+      freq, unit = FLAGS.eval_frequency_override.split(' ')
+      freq = int(freq)
+      assert (unit in ['epoch', 'step'])
+    except:
+      logging.error(
```

Comment:
Similar to above, instead of logging and raise, I would `raise ValueError('Failed to parse e...')`

Reply:
Done. Thanks!